fmt/mod.c

```diff
@@ -396,3 +468,216 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
+				mod_sampleheader[29] = (song->samples[n].loop_end - song->samples[n].loop_start) >> 1;// loop length LSB /2
+			} else {
+				mod_sampleheader[26] = 0;
+				mod_sampleheader[27] = 0;
+				mod_sampleheader[28] = 0;
+				mod_sampleheader[29] = 1;
+			}
+		}
+		disko_write(fp, mod_sampleheader, 30); // writing current sample header
+	}
+
+	tmp[0] = nord = csf_get_num_orders(song); // or "csf_get_num_orders(song_t *csf);"
+	tmp[1] = 0x7f;
+	disko_write(fp, tmp, 2);
+
+	for(maxpat = i = 0; (i < nord) && (i < 128); ++i) {
+		mod_orders[i] = song->orderlist[i];
+		if(maxpat < mod_orders[i]) maxpat = mod_orders[i];
+	}
+	for(; i < 128; ++i)
+		mod_orders[i] = 0;
+
+	disko_write(fp, mod_orders, 128);
+
+	if(maxpat < 64)
+		disko_write(fp, valid_tags[0][0], 4);
+	else
+		disko_write(fp, valid_tags[1][0], 4);
+
+	for(n = 0; n <= maxpat; ++n) {
+		m = song->patterns[n];
+		for(i = 0; i < 1024; ++i)
+			mod_pattern[i] = 0;
+		jmax = song->pattern_size[n];
+		if(jmax != 64) {
+			if(jmax > 64)
+				jmax = 64;
+			warn |= 1 << WARN_PATTERNLEN;
+		}
+		jmax *= MAX_CHANNELS;
+		for (j = joutpos = 0; j < jmax; ++j, ++m) {
+			uint8_t mod_fx, mod_fx_val;
+			if ((j % MAX_CHANNELS) < 4) {
+				//if (m->note || m->instrument || m->voleffect || m->volparam || m->effect || m->param) {...
+				period = amigaperiod_table[(m->note) & 0xff];
+				if (((m->note) & 0xff) && ((period < 113) || (period > 856)))
+					warn |= 1 << WARN_NOTERANGE;
+				mod_pattern[joutpos] = ((m->instrument) & 0x10) | (period >> 8);
+				mod_pattern[joutpos + 1] = period & 0xff;
+				mod_pattern[joutpos + 2] = (m->instrument & 0xf) << 4;
+				mod_fx = 0; mod_fx_val = m->param;
+				if(m->voleffect == VOLFX_VOLUME) {
+					mod_fx = 0x0c;
+					mod_fx_val = m->volparam;
+				} else if (m->voleffect == VOLFX_NONE) {
+					switch(m->effect) {
+						case FX_NONE: mod_fx_val = 0; break;
+						case FX_ARPEGGIO: mod_fx = 0; break;
+						case FX_PORTAMENTOUP:
+							mod_fx = 1;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_PORTAMENTODOWN:
+							mod_fx = 2;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_TONEPORTAMENTO: mod_fx = 3; break;
+						case FX_VIBRATO: mod_fx = 4; break;
+						case FX_TONEPORTAVOL: mod_fx = 5; break;
+						case FX_VIBRATOVOL: mod_fx = 6; break;
+						case FX_TREMOLO: mod_fx = 7; break;
+						case FX_PANNING: mod_fx = 8; break;
+						case FX_OFFSET: mod_fx = 9; break;
+						case FX_VOLUMESLIDE:
+							mod_fx = 0x0a;
+							if ((mod_fx_val & 0xf0) == 0xf0) { // fine volslide down!
```

Comment:
Note that DF0 and D0F are currently not handled correctly by this conversion: They are converted to EA0 and EB0, but should be converted to AF0 and A0F.

Reply:
I will push the fix in a moment.

Reply:
Done.